_posts/2019-12-20-differential-privacy/differential_privacy.Rmd

```diff
@@ -0,0 +1,470 @@
+
+
+Obviously, choice of $\epsilon$ is a (challenging) topic by itself, and not something we can elaborate on in a post dedicated to the technical aspects of DP with TensorFlow.
+
+How would $\epsilon$ change if we trained for 50 epochs instead? (This is actually what we'll do, seeing that training results on the test set tend to jump around quite a bit.)
+
+```{python}
+python compute_dp_sgd_privacy.py --N=3989 --batch_size=32 --noise_multiplier=1.1 --epochs=60
+```
+
+```
+DP-SGD with sampling rate = 0.802% and noise_multiplier = 1.1 iterated over
+6233 steps satisfies differential privacy with eps = 4.25 and delta = 1e-06.
+```
+Having talked about its parameters, now let's define the DP optimizer:
+
+```{r}
+l2_norm_clip <- 1
+noise_multiplier <- 1.1
+num_microbatches <- tf$cast(batch_size, tf$int32)
+learning_rate <- 0.01
+
+optimizer <- priv$DPAdamGaussianOptimizer(
+  l2_norm_clip = l2_norm_clip,
+  noise_multiplier = noise_multiplier,
+  num_microbatches = num_microbatches,
+  learning_rate = learning_rate
+)
+
+```
+
+One other change to make for DP. As gradients are clipped on a per-sample basis, the optimizer needs to work with per-sample losses as well: 
+
+```{r}
+loss <- tf$keras$losses$MeanSquaredError(reduction =  tf$keras$losses$Reduction$NONE)
+
+```
+
+Everything else stays the same. Training history (like we said above, lasting for 50 epochs now) looks a lot more turbulent, with MAEs on the test set fluctuating between 8 and 20 over the last 10 training epochs:
+
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset", fig.cap = "Training history with differential privacy."}
+knitr::include_graphics("images/dp.png")
+```
+
+
+In addition to the above-mentioned command line script, we can also compute $\epsilon$ as part of the training code. Let's double check:
+
+```{r}
+# probability of an individual training point being included in a minibatch
+sampling_probability <- batch_size / n_train
+
+# number of steps the optimizer takes over the training data
+steps <- num_epochs * n_train / batch_size
+
+# required for reasons related to how TF Privacy computes privacy
+# this actually is Renyi Differential Privacy: https://arxiv.org/abs/1702.07476
+# we don't go into details here and use same values as the command line script
+orders <- c((1 + (1:99)/10), 12:63)
+
+rdp <- priv$privacy$analysis$rdp_accountant$compute_rdp(
+  q = sampling_probability,
+  noise_multiplier = noise_multiplier,
+  steps = steps,
+  orders = orders)
+
+priv$privacy$analysis$rdp_accountant$get_privacy_spent(
+  orders, rdp, target_delta = 1e-6)[[1]]
+
+```
+
+```
+[1] 4.249645
+```
+
+So, we do get the same result. This was a long post; let's wrap up.
```

Comment:
```suggestion
So, we do get the same result.
```

Reply:
I like having that conclusion on its own, since it's quite a journey to get there. I think the `## Conclusion` header is sufficient for signalling that you're wrapping up. 😸

Reply:
ooooooh yeah... among every thing else, these are the kinds of comments I love your reviews for :-))) (take out the fill words / fill phrases or however that's called ... the no-information stuff.)